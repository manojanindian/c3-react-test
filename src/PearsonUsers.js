import React, { Component } from "react";

/**
 * UserProfile component will render following elements
 * User's image, name and delete button
 */

const UserProfile = ({first_name, last_name, avatar, id, onClickDeleteUser, position}) => {
	return (
		<div className="pearson-users__user">
			<img
				alt={`${first_name} ${last_name}`}
				className="pearson-users__user__avatar"
				src={avatar}
			/>
			<h3
				className="pearson-users__user__name"
			>
				{`${first_name} ${last_name}`}
			</h3>
			<button
				className="pearson-users__user__delete-btn"
				type="button"
				onClick={() => onClickDeleteUser(id, position)}
			>
				Delete
			</button>
		</div>
	)
}

export class PearsonUsers extends Component {
	constructor(props) {
		super(props);

		this.state = {
			users: [
				{
					id: 4,
					first_name: "Eve",
					last_name: "Holt",
					avatar:
					"https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
				},
				{
					id: 5,
					first_name: "Charles",
					last_name: "Morris",
					avatar:
					"https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
				},
				{
					id: 6,
					first_name: "Tracey",
					last_name: "Ramos",
					avatar:
					"https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
				}
			],
			isLoading: false
		};

		this.onClickDeleteUser = this.onClickDeleteUser.bind(this);
	}

	componentDidMount() {
		this.setState({
			isLoading: true
		});

		fetch('https://reqres.in/api/users?page=1&per_page=10')
		.then(data => data.json())
		.then((data) => {
			const newUserList = this.removeDuplicateEntries(data.data);

			this.setState((prevState, props) => {
				return {
					users: [...prevState.users, ...newUserList],
					isLoading: false
				};
			});
		});
  }

	/**
	 * removeDuplicateEntries() function is responsible to remove duplicate entries from fetch response
	 * function will remove duplicate entries and will concate oldUsers data with newUsers data
	 */

	removeDuplicateEntries(newUsersList) {
		const existingUsersId = new Set(this.state.users.map(({id}) => id));
		return newUsersList.filter(user => !existingUsersId.has(user.id));
	}

	/**
	 * onClickDeleteUser() function will remove respected user from component state
	 */

	onClickDeleteUser(userId, position) {
		this.setState((prevState, props) => {
			if(prevState.users[position].id === userId) {
				prevState.users.splice(position, 1)
			}
			return {
				users: prevState.users,
				isLoading: false
			};
		});
	}

	render() {
		const usersProfile = this.state.users.map( (user, index) =>
			<UserProfile
				key={index}
				{...user}
				position={index}
				onClickDeleteUser={this.onClickDeleteUser}
			/>
		);

		return (
			<div className="pearson-users">
				<h1 className="pearson-users__heading">
					Pearson User Management
				</h1>
				<div className="pearson-users__list">
					{usersProfile}
				</div>
				{this.state.isLoading &&
					<div className="pearson-users__loading">
						Loading...
					</div>
				}
			</div>
		);
	}
}
