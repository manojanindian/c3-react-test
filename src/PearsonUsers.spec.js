import React from "react";
import { shallow } from "enzyme";
import { PearsonUsers } from "./PearsonUsers";

describe("PearsonUsers", () => {
	let component;
	/**
	 * Mock global fetch funciton
	 * fetch will return response having 3 users
	 * In 3 users one is duplicate which id is 4
	 */
	
	beforeEach(() => {
		fetch = jest.fn().mockImplementationOnce(() => {
			return new Promise((resolve, reject) => {
				resolve({
					json: () => {
						return {data: [
							{
								id: 1,
								first_name: "Eve",
								last_name: "Holt",
								avatar:
									"https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
							},
							{
								id: 4,
								first_name: "Charles",
								last_name: "Morris",
								avatar:
									"https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
							},
							{
								id: 8,
								first_name: "Charles",
								last_name: "Morris",
								avatar:
									"https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
							}
						]};
					},
				});
			});
		});
	
		component = shallow(<PearsonUsers />);
		
	});

	it("component should be defined", () => {
		expect(component).toBeDefined();
	});

	it("component has heading", () => {
		const heading = component.find(".pearson-users__heading");
		expect(heading.text()).toEqual("Pearson User Management");
	});

	/**
	 * UserProfile componet must containe following three elements
	 * User avatar, user name and delete button
	 */
	  
	it("User Profile has user avatar", () => {
		const userProfileComponent = component.find('UserProfile').first().dive();
		expect(userProfileComponent.find('.pearson-users__user__avatar').length).toBe(1);
	});

	it("User Profile has user name", () => {
		const userProfileComponent = component.find('UserProfile').first().dive();
		expect(userProfileComponent.find('.pearson-users__user__name').length).toBe(1);
	});

	it("User Profile has delete btn", () => {
		const userProfileComponent = component.find('UserProfile').first().dive();
		expect(userProfileComponent.find('.pearson-users__user__delete-btn').length).toBe(1);
	});

	it("fetch should be called", () => {
		expect(fetch.mock.calls.length).toBe(1);
	});

	it("fetch url should be 'https://reqres.in/api/users?page=1&per_page=10'", () => {
		expect(fetch.mock.calls[0][0]).toBe('https://reqres.in/api/users?page=1&per_page=10');
	});

	/**
	 * PearsonUsers component has already 3 users data in state [4, 5, 6]
	 * Fetch is providing 3 more users data [1, 4, 8]
	 * Total users should be 6 now but we have functionality to remove duplicate users [4]
	 * hence Total users should be 5 instead 6
	 */
	
	it("duplicate user should be deleted", () => {
		expect(component.state().users.length).toBe(5);
	});

	/**
	 * After fetch PersonUsers component has 5 users in state [4, 5, 6, 1, 8]
	 * If user click on delete button user's data should be deleted from state
	 * After delete total number of users will be 4 in state
	 */
	
	it("delete user onclick of delete button", () => {
		const userProfileComponent = component.find('UserProfile').first().dive();
		userProfileComponent.find('.pearson-users__user__delete-btn').simulate('click');
		expect(component.state().users.length).toBe(4);
	});

	it("component should render correctly", () => {
		expect(component).toMatchSnapshot();
	})
});
