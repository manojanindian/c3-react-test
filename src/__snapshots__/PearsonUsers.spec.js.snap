// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PearsonUsers component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PearsonUsers />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="pearson-users__heading"
>
          Pearson User Management
</h1>,
        <div
          className="pearson-users__list"
>
          <UserProfile
                    avatar="https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
                    first_name="Eve"
                    id={4}
                    last_name="Holt"
                    onClickDeleteUser={[Function]}
                    position={0}
          />
          <UserProfile
                    avatar="https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
                    first_name="Charles"
                    id={5}
                    last_name="Morris"
                    onClickDeleteUser={[Function]}
                    position={1}
          />
          <UserProfile
                    avatar="https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
                    first_name="Tracey"
                    id={6}
                    last_name="Ramos"
                    onClickDeleteUser={[Function]}
                    position={2}
          />
</div>,
        <div
          className="pearson-users__loading"
>
          Loading...
</div>,
      ],
      "className": "pearson-users",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Pearson User Management",
          "className": "pearson-users__heading",
        },
        "ref": null,
        "rendered": "Pearson User Management",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <UserProfile
              avatar="https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
              first_name="Eve"
              id={4}
              last_name="Holt"
              onClickDeleteUser={[Function]}
              position={0}
/>,
            <UserProfile
              avatar="https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
              first_name="Charles"
              id={5}
              last_name="Morris"
              onClickDeleteUser={[Function]}
              position={1}
/>,
            <UserProfile
              avatar="https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
              first_name="Tracey"
              id={6}
              last_name="Ramos"
              onClickDeleteUser={[Function]}
              position={2}
/>,
          ],
          "className": "pearson-users__list",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "function",
            "props": Object {
              "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
              "first_name": "Eve",
              "id": 4,
              "last_name": "Holt",
              "onClickDeleteUser": [Function],
              "position": 0,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "function",
            "props": Object {
              "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
              "first_name": "Charles",
              "id": 5,
              "last_name": "Morris",
              "onClickDeleteUser": [Function],
              "position": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "function",
            "props": Object {
              "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
              "first_name": "Tracey",
              "id": 6,
              "last_name": "Ramos",
              "onClickDeleteUser": [Function],
              "position": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Loading...",
          "className": "pearson-users__loading",
        },
        "ref": null,
        "rendered": "Loading...",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="pearson-users__heading"
>
            Pearson User Management
</h1>,
          <div
            className="pearson-users__list"
>
            <UserProfile
                        avatar="https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
                        first_name="Eve"
                        id={4}
                        last_name="Holt"
                        onClickDeleteUser={[Function]}
                        position={0}
            />
            <UserProfile
                        avatar="https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
                        first_name="Charles"
                        id={5}
                        last_name="Morris"
                        onClickDeleteUser={[Function]}
                        position={1}
            />
            <UserProfile
                        avatar="https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
                        first_name="Tracey"
                        id={6}
                        last_name="Ramos"
                        onClickDeleteUser={[Function]}
                        position={2}
            />
</div>,
          <div
            className="pearson-users__loading"
>
            Loading...
</div>,
        ],
        "className": "pearson-users",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Pearson User Management",
            "className": "pearson-users__heading",
          },
          "ref": null,
          "rendered": "Pearson User Management",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <UserProfile
                avatar="https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
                first_name="Eve"
                id={4}
                last_name="Holt"
                onClickDeleteUser={[Function]}
                position={0}
/>,
              <UserProfile
                avatar="https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
                first_name="Charles"
                id={5}
                last_name="Morris"
                onClickDeleteUser={[Function]}
                position={1}
/>,
              <UserProfile
                avatar="https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
                first_name="Tracey"
                id={6}
                last_name="Ramos"
                onClickDeleteUser={[Function]}
                position={2}
/>,
            ],
            "className": "pearson-users__list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "function",
              "props": Object {
                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
                "first_name": "Eve",
                "id": 4,
                "last_name": "Holt",
                "onClickDeleteUser": [Function],
                "position": 0,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "function",
              "props": Object {
                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
                "first_name": "Charles",
                "id": 5,
                "last_name": "Morris",
                "onClickDeleteUser": [Function],
                "position": 1,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "function",
              "props": Object {
                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
                "first_name": "Tracey",
                "id": 6,
                "last_name": "Ramos",
                "onClickDeleteUser": [Function],
                "position": 2,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Loading...",
            "className": "pearson-users__loading",
          },
          "ref": null,
          "rendered": "Loading...",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
